integrationTests/vm/arwen/arwenvm/deployment/upgrade_test.go

```diff
@@ -0,0 +1,105 @@
+package deployment
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go-core/data/block"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm"
+	"github.com/ElrondNetwork/elrond-go/integrationTests/vm/arwen"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestScUpgradeShouldManageCorrectlyTheCodeMetadata(t *testing.T) {
+	senderAddressBytes := []byte("12345678901234567890123456789012")
+	senderNonce := uint64(0)
+	senderBalance := big.NewInt(100000000)
+
+	testContext, err := vm.CreatePreparedTxProcessorAndAccountsWithVMs(
+		senderNonce,
+		senderAddressBytes,
+		senderBalance,
+		config.EnableEpochs{
+			IsPayableBySCEnableEpoch: 1,
+		},
+	)
+	require.Nil(t, err)
+	defer testContext.Close()
+
+	t.Run("payable by SC is not active, code metadata is not parsed", func(t *testing.T) {
+		testContext.EpochNotifier.CheckEpoch(&block.Header{Epoch: 0})
+
+		contractAddress := deployDummySCReturningContractAddress(t, testContext, senderAddressBytes, "0502")
+
+		expectedCodeMetadata := vmcommon.CodeMetadata{
+			Payable:     true,
+			PayableBySC: false,
+			Upgradeable: true,
+			Readable:    true,
+		}
+
+		assert.Equal(t, expectedCodeMetadata.ToBytes(), getCodeMetadata(t, testContext.Accounts, contractAddress))
+
+		newCodeMetadata := []byte{0xFF, 0xFF, 0xFF}
+		upgradeDummySCReturningContractAddress(t, testContext, senderAddressBytes, contractAddress, hex.EncodeToString(newCodeMetadata))
+
+		assert.Equal(t, newCodeMetadata, getCodeMetadata(t, testContext.Accounts, contractAddress))
+	})
+	t.Run("payable by SC is active, code metadata is parsed", func(t *testing.T) {
```

Comment:
could there be also the case with payable by SC active, but code metadata not ok: FFFFFF? Could this change the contractMetadata to 0000 and make it not upgradeable, not payable etc like on the deploy?

Reply:
This is quite severe, I would opt also for returning an error here. What do you think, @sasurobert ?

Reply:
done for both direct and indirect deployments